### ISSUE TITLE
typescript-best-practices/no-empty-pattern: false positives on empty array/object literals (rule matches node type without checking binding position)

### ISSUE THREAD

**Issue by SebastianBienert** [DataDog/datadog-static-analyzer#961]
**Describe the bug**

`typescript-best-practices/no-empty-pattern` reports `Unexpected empty array pattern.` / `Unexpected empty object pattern.` on empty array and object **literals**, not on destructuring patterns.

On one 962-line file of valid TypeScript it produces 11 findings, all false. Across our Angular/Nx repo it produced 28 findings in 8 files, all false. Examples of flagged code:

```ts
return input ?? [];                                   // Unexpected empty array pattern.
const entities = cloneDeep(state.entities) ?? {};     // Unexpected empty object pattern.
const {a, b} = config ?? {};                          // Unexpected empty object pattern.
const filters: Filter[] = [];                         // Unexpected empty array pattern.
```

The SARIF region points exactly at the `[]` / `{}` token. The [rule documentation](https://docs.datadoghq.com/security/code_security/static_analysis/static_analysis_rules/typescript-best-practices/no-empty-pattern/) lists these forms as **compliant** (`var {a = []} = foo;`), so the behavior contradicts the documented intent.

**Probable cause**

The rule, as served by `/api/v2/static-analysis/rulesets/typescript-best-practices`, matches on node type only:

```
[
    (object_pattern) @object_pattern
    (array_pattern) @array_pattern
]
```

```js
function visit(query, filename, code) {
  const objectPattern = query.captures["object_pattern"];
  const arrayPattern = query.captures["array_pattern"];
  const node = objectPattern || arrayPattern;
  const type = objectPattern ? 'object' : 'array';
  if (!node.children.length) {
    addError(buildError(node.start.line, node.start.col, node.end.line, node.end.col,
      `Unexpected empty ${type} pattern.`, 'ERROR', 'BEST_PRACTICES'))
  }
}
```

`tree-sitter-javascript` declares literal/pattern ambiguity as GLR conflicts ([grammar.js](https://github.com/tree-sitter/tree-sitter-javascript/blob/master/grammar.js)):

```js
[$.array, $.array_pattern],
[$.object, $.object_pattern],
```

So an empty `[]` or `{}` can be typed `array_pattern` / `object_pattern` depending on parse context. Because the rule never checks that the capture sits in a **binding position**, a literal typed as a pattern is reported as an empty destructuring pattern.

This also explains the distribution: the same idiom occurs ~2100 times in our `apps/` and `libs/`, and only 28 are flagged, clustered in a few files and regions (12 in one file), with adjacent identical statements differing:

```ts
let values: string[] | undefined = [];       // flagged
let valueRanges: ValueRange[] | undefined = [];  // not flagged
```

**To Reproduce**

A short snippet does not reproduce it, precisely because the trigger depends on parse context. This 60-line file does, deterministically. It parses with **0 diagnostics** via `ts.createSourceFile` (TypeScript 6.0.3), and was produced by delta-debugging the original file while requiring syntactic validity at every step:

```ts
export class Helper {
  private static buildAll(
  ): Bundle {
    if (listA.length > 0) {
    }
    const rows = listB.map((f) => {
      return {
      };
    });
    if (filter.parameter === CATEGORY_A) {
      if (mapA && keyA in mapA) {
        for (const valueA of values) {
        }
      }
    }
    config.items.forEach((item) => {
    });
  }
  public static normalizeItems(filters: Item[]): Item[] {
    return filters?.map((filter: Item) => {
      return Object.keys(filter)
        .filter((k: string) => !isNil(filter[<keyof Item>k]) && !isEmpty(filter[<keyof Item>k]))
    });
  }
  public static getKey(
  ): string {
    return this.buildAll(
    );
  }
  public static normalizeBundle(
  ): Bundle {
    return {
      fieldA: input?.fieldA?.map(
          ({
          }) as TypeA,
      ),
    };
  }
  public static fromBundle(
  ): Item[] {
    for (const itemD of bundle.listD ?? []) {
      if (flagA || flagB) {
      }
    }
    if (bundle.listE?.length) {
    }
    if (!isEmpty(mapB)) {
      for (const itemB of Object.values(mapB)) {
      }
    }
    if (bundle.flagC) {
    }
    return rows?.map((r) => {
      const {id, level, label: name, child, iso: country, parent, parentId: parentId} = r ?? {};
      const icon = ICON_A;
      return {
      } as Option;
    });
  }
}
```

1. Save as `candidate.ts` in an otherwise empty directory, alongside:

```yaml
schema-version: v1.4
sast:
  use-default-rulesets: false
  use-rulesets:
    - typescript-best-practices
```

2. Run `datadog-static-analyzer -i . -o out.sarif -f sarif`
3. Result:

```
typescript-best-practices/no-empty-pattern line 54 cols 94-96
```

Line 54 is `const {id, level, label: name, child, iso: country, parent, parentId: parentId} = r ?? {};`, and columns 94-96 are the `{}` of the `??` default — an object literal used as a fallback value, not a destructuring pattern. The non-empty pattern on that line's left side is correct code and is not what gets reported.

(The reducer also leaves empty blocks behind, so the same run reports `typescript-best-practices/no-empty` at lines 4 and 12. Those are legitimate; only the `no-empty-pattern` finding is the bug.)

**Expected behavior**

No `no-empty-pattern` finding. Only genuine empty destructuring patterns should be reported, as the rule's documentation and test fixtures describe:

```ts
var {} = foo
var [] = foo
function foo({}) {}
```

Cross-check: ESLint's `no-empty-pattern` is enabled at error severity in the same repo (`eslint --print-config` confirms `"no-empty-pattern": [2, {...}]`) and reports **zero** problems on every file the Datadog rule flags.

**Suggested fix**

Gate the report on the capture being in a binding position, for example requiring the parent node to be one of `variable_declarator` (as the `name` field), `formal_parameters` / `required_parameter` / `optional_parameter`, the left side of an `assignment_expression`, a `for_in_statement` left side, or a `catch_clause` parameter. That keeps the documented cases and drops literals that the GLR parse happened to type as patterns.

**Additional context: `typescript-best-practices/no-extra-non-null-assertion` looks similar**

Same repo, 4 findings, all false. One is:

```ts
return !!versions && versions?.length > 1;   // Forbidden extra non-null assertion.
```

`!!` here is boolean coercion, not a duplicated non-null assertion. Two more findings point at a plain method signature with no `!` on the line at all. That rule's query includes anonymous-parent sibling sequences:

```
(((non_null_expression) (optional_chain)) @expression)
(((non_null_expression) "?.") @expression)
```

and its `visit` reports unconditionally on any capture, so it appears to share the pattern of trusting node classification without positional validation. Happy to open a separate issue if you prefer them tracked apart.

**Environment**

- `datadog-static-analyzer` 0.9.1, `x86_64-pc-windows-msvc` (installed via mise, `github:DataDog/datadog-static-analyzer`)
- OS: Windows 11
- `DD_SITE=datadoghq.eu`; ruleset `typescript-best-practices`, rule `no-empty-pattern`, checksum `415af295a8db1a33b56c3cf4839fa35874b497f2e333b374855796664722`, `last_updated_at` 2024-09-05
- Source project: Angular 22 / Nx 23 monorepo, TypeScript 6.0.3

Workaround in use: both rules disabled repo-wide via `ruleset-configs` → `rule-configs` → `ignore-paths: ["**"]`, since the false positives track parse context rather than specific source lines and so cannot be suppressed reliably per line or per file.

**Comment by SebastianBienert** [comment]
Filed the `no-extra-non-null-assertion` part separately as #962, since it is a different rule with a different query and root cause. This issue stays scoped to `no-empty-pattern`.

**Comment by jasonforal** [comment]
Hi, I appreciate the detailed description and reproduction! We'll take a look into this next week.

**Comment by jasonforal** [comment]
Hello, looks like this is due to how we parse `tsx` and `ts` with the same grammar. A fix for this has some nuance that we'd need to think through. To set expectations: we won't be able to address this in the short-term, but this is something we do want to fix.
